exchanges/stream/websocket.go

```diff
@@ -0,0 +1,921 @@
+		return fmt.Errorf("%s websocket: cannot subscribe no channels supplied",
+			w.exchangeName)
+	}
+	w.subscriptionMutex.Lock()
+	defer w.subscriptionMutex.Unlock()
+	for x := range channels {
+		for y := range w.subscriptions {
+			if channels[x].Equal(&w.subscriptions[y]) {
+				return fmt.Errorf("%s websocket: %v already subscribed",
+					w.exchangeName,
+					channels[x])
+			}
+		}
+	}
+	return w.Subscriber(channels)
+}
+
+// AddSuccessfulSubscriptions adds subscriptions to the subscription lists that
+// has been successfully subscribed
+func (w *Websocket) AddSuccessfulSubscriptions(channels ...ChannelSubscription) {
+	w.subscriptions = append(w.subscriptions, channels...)
+}
+
+// RemoveSuccessfulUnsubscriptions removes subscriptions from the subscription
+// list that has been successfulling unsubscribed
+func (w *Websocket) RemoveSuccessfulUnsubscriptions(channels ...ChannelSubscription) {
+	for x := range channels {
+		for y := range w.subscriptions {
+			if channels[x].Equal(&w.subscriptions[y]) {
+				w.subscriptions[y] = w.subscriptions[len(w.subscriptions)-1]
+				w.subscriptions[len(w.subscriptions)-1] = ChannelSubscription{}
+				w.subscriptions = w.subscriptions[:len(w.subscriptions)-1]
+				break
+			}
+		}
+	}
+}
+
+// Equal two WebsocketChannelSubscription to determine equality
+func (w *ChannelSubscription) Equal(s *ChannelSubscription) bool {
+	return strings.EqualFold(w.Channel, s.Channel) &&
+		w.Currency.Equal(s.Currency)
+}
+
+// GetSubscriptions returns a copied list of subscriptions
+// subscriptions is a private member and cannot be manipulated
+func (w *Websocket) GetSubscriptions() []ChannelSubscription {
+	return append(w.subscriptions[:0:0], w.subscriptions...)
+}
+
+// SetCanUseAuthenticatedEndpoints sets canUseAuthenticatedEndpoints val in
+// a thread safe manner
+func (w *Websocket) SetCanUseAuthenticatedEndpoints(val bool) {
+	w.subscriptionMutex.Lock()
+	defer w.subscriptionMutex.Unlock()
+	w.canUseAuthenticatedEndpoints = val
+}
+
+// CanUseAuthenticatedEndpoints gets canUseAuthenticatedEndpoints val in
+// a thread safe manner
+func (w *Websocket) CanUseAuthenticatedEndpoints() bool {
+	w.subscriptionMutex.Lock()
+	defer w.subscriptionMutex.Unlock()
+	return w.canUseAuthenticatedEndpoints
+}
+
+// isDisconnectionError Determines if the error sent over chan ReadMessageErrors is a disconnection error
+func isDisconnectionError(err error) bool {
+	if websocket.IsUnexpectedCloseError(err) {
+		return true
+	}
+	switch e := err.(type) {
+	case *websocket.CloseError:
+		return true
+	case *net.OpError:
+		if e.Err.Error() == "use of closed network connection" {
+			return false
+		}
+		return true
+	}
+	return false
+}
+
+// checkWebsocketURL checks for a valid websocket url
+func checkWebsocketURL(s string) error {
+	split := strings.Split(s, "://")
+	if len(split) < 1 || split[0] != "wss" && split[0] != "ws" {
+		return fmt.Errorf("cannot set invalid websocket URL %s", s)
+	}
+	return nil
```

Comment:
can use the Parse function from the URL package if you wish instead

```suggestion
	u, err := url.Parse(s)
	if err != nil {
		return err
	}
	if u.Scheme != "ws" && u.Scheme != "wss" {
		return fmt.Errorf("cannot set invalid websocket URL %s", s)
	}
        return nil
```